### PR TITLE
sc-398/poetry st

### PR DESCRIPTION
**Pull request by epcAnker**
Special unique things i had to do:
**Versioning** - was done in setup.py with versioneer.  I moved this out and did some of the versioning in the toml file with [tool.poetry-dynamic-versioning] (using the same format).  I then made the file "generate_versioning.py" to generate the _version.py file that was originally there.

buildui - the ui files are compiled in setup.py.  I created a file "buildui.py" to do the same thing.  I added this to both ST & STlib .  This needed to be added to the readme for manual instructions and to the ci

Removed all romp items, requirements, boots, setup.py

all venv -> .venv (poetry uses .venv)

------

To use:

first install poetry  
for powershell: `(Invoke-WebRequest -Uri https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py -UseBasicParsing).Content | python -`

linux: `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -`

`poetry install` should install dependencies and create scripts

`poetry run epyq`
 

---

(will put a poetry guide on confluence)

If you want the .venv to install in your directory:

`poetry config virtualenvs.in-project = true`  
`poetry config virtualenvs.path = "."`

`poetry config --list` shows all config.  mine looks like:

cache-dir = "C:\\\\Users\\\\Alex\\\\AppData\\\\Local\\\\pypoetry\\\\Cache"  
experimental.new-installer = true  
installer.parallel = true  
virtualenvs.create = true  
virtualenvs.in-project = true  
virtualenvs.path = "."

### CODE REVIEW

File: buildui.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        directory_paths=[pathlib.Path(__file__).parent / "src" / "epyq"],
+    )
+
+    # todo, both of these are the same.  change it only have one
+    alqtendpy.compileui.compile_ui(
+        directory_paths=[pathlib.Path(__file__).parent / "sub" / "epyqlib" / "epyqlib"],
+    )
+
+
+#    epyqlib.buildui.compile_ui()
```
Comment (epcAnker) [inline buildui.py:19]: which to call?  up for discussion

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -38,13 +49,12 @@ jobs:
...
-          venv/Scripts/black --check --diff .
-          venv/Scripts/python download_fonts.py src/libs
-          venv/Scripts/genbuildinfo "src/epyq/_build.py"
-          venv/Scripts/pip freeze --all
+          poetry run buildui
+          poetry run python ./generate_versioning.py
+          .venv/Scripts/black --exclude=.\.venv --check --diff .
+          .venv/Scripts/python download_fonts.py src/libs
+          .venv/Scripts/genbuildinfo "src/epyq/_build.py"
+          .venv/Scripts/pip freeze --all
```
Comment (epcAnker) [inline .github/workflows/ci.yml:57]: for some reason `poetry run ...` didnt work when i changed them all to that so i kept .venv/ instead

File: README.md [hunk 3]
```diff
@@ -40,14 +41,15 @@ EPyQ widgets should be visible at the bottom of the widget box on the left.
 - Install Python 3.7
   - Consider [pyenv](https://github.com/pyenv/pyenv) to get Python versions
 - Install git
-- `git clone https://github.com/altendky/st`
+- `git clone https://github.com/epcpower/st`
 - `cd st`
 - `git submodule update --init`
-- `python boots.py ensure`
+- `poetry install`
+- `poetry run buildui`
```
Comment (epcAnker) [inline README.md:48]: i am adding another manual step that was initially called in the setup.py.  not sure how to get this to run automatically.

File: buildui.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import alqtendpy.compileui
+import pathlib
+
```
Comment (epcAnker) [inline buildui.py:3]: Is the name builldui ok?  This builds the ui for both st and stlib.

File: tox.ini [hunk 5]
```diff
@@ -22,7 +22,7 @@ commands =
     "{envpython}" -m pytest -vvvv
 
 [testenv:devenv]
-envdir = {toxinidir}/venv
+envdir = {toxinidir}/.venv
```
Comment (epcAnker) [inline tox.ini:25]: not exactly sure what this file is doing at this point.

we use poetry build now which i believe uses tox.

File: buildui.py (hunk 4 above)
Comment (gordon-epc) [reply buildui.py:3]: The name is ok. If I was going to suggest something else, I'd put underscores in there. `build_ui.py`. Not sure if it's worth it at this point.

File: src/epyq/_build.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# This file should be overwritten when a distribution build is done
+# This file has been generated
```
Comment (gordon-epc) [inline src/epyq/_build.py:1]: Do we need to change this comment? I think the previous comment is more appropriate. With the new comment, it almost seems like the file was accidentally checked in.

File: .gitignore [hunk 7]
```diff
@@ -1,4 +1,5 @@
 /venv/
+/.venv/
```
Comment (gordon-epc) [inline .gitignore:2]: I think the `.gitignore` could be autogenerated with `gitignoreio` like most of our other repos. I'll do a PR for it.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

**Comment by gordon-epc** [comment]
Another thing I noticed. The `designer.bat` file is not available. This allows a developer to run the PyQt designer application with the EPyQ specific widgets available. I'm not exactly sure how `designer.bat` is created (maybe `pyqt-tools`)? It's important, though. Developers can't edit the UI properly without it.

File: README.md [hunk 8]
```diff
@@ -11,27 +11,25 @@ Most of the interesting parts of the GUI are loaded from device configuration fi
...
 - Note the links provided for possibly needed system and driver extras.
-- Try running `venv\Scripts\epyq`.  If you get errors, consider installing the linked extras. 
+- Try running `poetry run epyq`.  If you get errors, consider installing the linked extras.
 
 If using with [PEAK PCAN](http://www.peak-system.com/PCAN-USB.199.0.html?&L=1) hardware, install the [PEAK drivers](http://www.peak-system.com/PCAN-USB.199.0.html?&L=1).
 Select the `PCAN-Basic` feature in the installer.  The virtual and LIN features are not needed for this.
 
-To launch EPyQ run `venv\Scripts\epyq.exe`.
+To launch EPyQ run `poetry run epyq`.
 To launch Qt Designer with the EPyQ plugins enabled run `designer.bat`.
```
Comment (gordon-epc) [inline README.md:33]: The file `designer.bat` is not created with a poetry install. It's not clear where this file actually originates.

Comment (gordon-epc) [inline README.md:33]: The file `designer.bat` is not created with a poetry install. It's not clear where this file actually originates.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

**Review by gordon-epc — CHANGES_REQUESTED** [review]

**Review by bholewin-epc — APPROVED** [review]
Approved but please wait to merge until Gordon signs off as well.

File: src/epyq/_build.py [hunk 9]
```diff
@@ -1,6 +1,5 @@
 # This file should be overwritten when a distribution build is done
 
-
```
Comment (gordon-epc) [inline src/epyq/_build.py:3]: Nitpick. This is the only change in the file and it's not necessary. Undo, please.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: src/epyq/_build.py (hunk 9 above)
Comment (epcAnker) [reply src/epyq/_build.py]: 6f19bd2365857e95228cba5bbde21aa4c3f41845

File: pyproject.toml [hunk 10]
```diff
@@ -1,19 +1,97 @@
+[tool.poetry]
+name = "epyq"
+version = "0.0.0"
+description = ""
```
Comment (gordon-epc) [inline pyproject.toml:4]:
```suggestion
description = "EPyQ (EPC Service Tool)"
```

File: pyproject.toml [hunk 11]
```diff
@@ -1,19 +1,97 @@
...
+authors = ["alex anker <alex.anker@epcpower.com>"]
+license = "GPL-2.0-or-later"
+
+[tool.poetry-dynamic-versioning]
+enable = true
```
Comment (gordon-epc) [inline pyproject.toml:9]: I see this set as `enable = false` elsewhere. Is this correct?

File: pyproject.toml [hunk 12]
```diff
@@ -1,19 +1,97 @@
...
+Twisted = "21.2.0"
+qt5reactor = "0.5"
+pyelftools = "0.26"
+attrs = ">=19.3.0"
+natsort = "5.5.0"
+arrow = "0.12.1"
+PyQtChart = "5.14.0"
+pywin32 = { version = "223", markers = "sys_platform == 'win32'" }
+certitude = { version = "1.0.1", markers = "sys_platform == 'Darwin'" }
+pytest-twisted = "<=1.13.4"
```
Comment (gordon-epc) [inline pyproject.toml:47]: Try `pytest-twisted` in `tool.poetry.dev-dependencies`?

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 11 above)
Comment (epcAnker) [reply pyproject.toml:9]: I think i had it working for dev setups. So poetry-dynamic-versioning would trigger for some poetry commands.  Ill set it to false though to be less confusing.

Comment (epcAnker) [reply pyproject.toml:9]: I think i had it working for dev setups. So poetry-dynamic-versioning would trigger for some poetry commands.  Ill set it to false though to be less confusing.

**Review by gordon-epc — APPROVED** [review]

**Pull request merged**